Index: b/src/firewall.c
===================================================================
--- a/src/firewall.c	2017-10-16 19:18:52.000000000 +0800
+++ b/src/firewall.c	2017-11-09 10:59:57.116197718 +0800
@@ -77,15 +77,15 @@ fw_allow(t_client * client, int new_fw_c
     debug(LOG_DEBUG, "Allowing %s %s with fw_connection_state %d", client->ip, client->mac, new_fw_connection_state);
     client->fw_connection_state = new_fw_connection_state;
 
-    /* Grant first */
-    result = iptables_fw_access(FW_ACCESS_ALLOW, client->ip, client->mac, new_fw_connection_state);
-
-    /* Deny after if needed. */
+    /* Deny if needed. */
     if (old_state != FW_MARK_NONE) {
         debug(LOG_DEBUG, "Clearing previous fw_connection_state %d", old_state);
         _fw_deny_raw(client->ip, client->mac, old_state);
     }
 
+    /* Grant first */
+    result = iptables_fw_access(FW_ACCESS_ALLOW, client->ip, client->mac, new_fw_connection_state);
+
     return result;
 }
 
