Index: b/src/gateway.c
===================================================================
--- a/src/gateway.c	2017-12-09 19:03:25.777024894 +0800
+++ b/src/gateway.c	2017-12-09 19:03:58.685161561 +0800
@@ -463,7 +463,7 @@ int get_config_from_server(const char *m
     pos = 0;
     len = 0;
     do{
-        len = recv(sock_cli, buffer + pos, sizeof(buffer), 0);
+        len = recv(sock_cli, buffer + pos, sizeof(buffer) - pos - 1, 0);
         pos += len;
     }while((len > 0) && (pos < sizeof(buffer) - 1));
     buffer[pos] = 0;
